### DISCUSSION TITLE
[How To] Update ioBroker js-controller in Portainer

### DISCUSSION THREAD

**Discussion by gyroslav** [buanet/ioBroker.docker#193] [Share your Experience / Teile deine Erfahrung]
hello world,
here a small HowTo update your iobroker in docker through portainer:

1. i assume, you have portainer running

2. navigate to your running iobroker-container and click "console"
![grafik](https://user-images.githubusercontent.com/79666110/135080990-a866a640-e426-4add-97b2-409f6dd3772f.png)

3. click CONNECT

4. to stop the iobroker daemon: 
`pkill -u iobroker`
google on how to stop iobroker will give you mostly "pkill io" and this will close the terminal und restart iobroker.
so the flag "-u" is very importand here.

5. from here you can follow the official instructions:
`iobroker update`
`iobroker upgrade self`
`iobroker start`

6. restart the container

**Comment by fzzybllz** [comment]
Hi @gyroslav,

very nice idea, but unfortunately your how to is old and i guess deprecated, isn't it @buanet?

Steps 1 to 3 are still ok, but step 4 is fixed by the maintenance script. So step 4 has to be `maintenance on` -> confirm with yes
Step 5 is partially fine, but instead of `iobroker start` you have to switch the maintenance script off by `maintenance off` + confirmation.
Step 6 will be performed automatically by the maintenance script

**Reply by gyroslav** [reply]
Actually the official instructions you get, after clicking on the Update-Button located here:
![grafik](https://user-images.githubusercontent.com/79666110/135261387-0d0e8924-87a0-4b5f-abe4-c3d26f58ed7c.png)
![grafik](https://user-images.githubusercontent.com/79666110/135261446-b7df451e-e1bb-4ae0-9379-3f144055544f.png)

are:

`iobroker stop`
`iobroker update`
`iobroker upgrade self`
`iobroker start`

i cannot show it right now, because it is allready done :)

**Reply by fzzybllz** [reply]
That's right, but it's meant to be for a plain installation of iobroker without docker. Iobroker does not detect, if it's running in docker or directly on the OS.

**Reply by gyroslav** [reply]
OK next time there is an update, I will look for the "maintenance on" instructions. And change my howto above.

**Reply by McTristan** [reply]
I'm quite a bit confused of how the iobroker docker image update seems to work when persisting the volume information - not at all it seems as the mapped volume contains the whole iobroker installation.

It seems I'm running V5 of iobroker at the moment (even though I'm running the latest image which is V6 based).

With the above described method I can at least update to v5.2.3 but what would I need to get the latest version 6?

Or would it be wise to ditch the mapped volume, start from scratch und restore from a backup?

Or is there an "easy" way to upgrade to v6?

**Comment by gyroslav** [comment]
Thats weird, i get only the Versions 3.3.22
![grafik](https://user-images.githubusercontent.com/79666110/147876846-747cd458-acf9-43a1-8cb4-1cef1a37cef6.png)

I will use this instruction for my next update: https://smarthome.buanet.de/2020/10/iobroker-docker-container-updates-upgrades/

**Comment by McTristan** [comment]
Actually I meant the iobroker version and not the JS Controller version. As the newest Docker image is based on v6.x I was wondering how to archive the same when using a volume map as suggested by the maintainer. When I update the container either by using the above mentioned method (using the iobroker commandline tools) or by using the newest image (I'm using portainer to update my containers) I always end up with a v5 version of iobroker.
So either I'm doing something wrong or there must be another way of doing this...

**Reply by McTristan** [reply]
To be more specific:

iobroker update

gives me the following list:

```
Adapter    "admin"         : 5.2.3    , installed 5.2.3
Adapter    "alexa2"        : 3.11.2   , installed 3.11.2
Adapter    "awattar"       : 1.0.6    , installed 1.0.6
Adapter    "backitup"      : 2.2.2    , installed 2.2.2
Adapter    "daswetter"     : 3.0.9    , installed 3.0.9
Adapter    "discovergy"    : 0.5.8    , installed 0.5.8
Adapter    "discovery"     : 2.7.3    , installed 2.7.3
Adapter    "feiertage"     : 1.0.23   , installed 1.0.23
Adapter    "fritzbox"      : 0.2.1    , installed 0.2.1
Adapter    "fritzdect"     : 2.1.15   , installed 2.1.15
Adapter    "harmony"       : 1.2.2    , installed 1.2.2
Adapter    "hmip"          : 1.13.2   , installed 1.13.2
Adapter    "hue-extended"  : 2.0.0    , installed 2.0.0
Adapter    "influxdb"      : 1.9.5    , installed 1.9.5
Adapter    "info"          : 1.9.8    , installed 1.9.8
Adapter    "javascript"    : 5.2.13   , installed 5.2.13
Controller "js-controller" : 3.3.22   , installed 3.3.22
Adapter    "lgtv"          : 1.1.10   , installed 1.1.10
Adapter    "lightify"      : 0.2.16   , installed 0.2.16
Adapter    "mihome-vacuum" : 3.3.1    , installed 3.3.1
Adapter    "mqtt"          : 2.4.0    , installed 2.4.0
Adapter    "node-red"      : 2.4.1    , installed 2.4.1
Adapter    "nuki-extended" : 2.3.1    , installed 2.3.1
Adapter    "nut"           : 1.5.0    , installed 1.5.0
Adapter    "octoprint"     : 1.1.1    , installed 1.1.1
Adapter    "parser"        : 1.0.7    , installed 1.0.7
Adapter    "pushover"      : 2.0.5    , installed 2.0.5
Adapter    "simple-api"    : 2.6.1    , installed 2.6.1
Adapter    "socketio"      : 3.1.4    , installed 3.1.4
Adapter    "sonoff"        : 2.4.5    , installed 2.4.5
Adapter    "sonos"         : 2.1.1    , installed 2.1.1
Adapter    "tradfri"       : 3.0.1    , installed 3.0.1
Adapter    "tuya"          : 3.6.9    , installed 3.6.9
Adapter    "virtualpowermeter": 1.4.2   , installed 1.4.2
Adapter    "web"           : 3.4.9    , installed 3.4.9
Adapter    "wled"          : 0.6.3    , installed 0.6.3
```

iobroker upgrade does effectively nothing as everything is already installed (v5.x -> admin module).
My docker image is already the latest version and apt-get update / upgrade do not produce any changes. Node 14.x is already installed.

**Reply by buanet** [reply]
> I always end up with a v5 version of iobroker.

How do you check the version of your ioBroker Docker container? 

Attention: The version of the Docker Image has nothing to do with any version of the ioBroker components like adapters or js-controller! Also the labels of a container, as you can see them in Portainer, may not be updated when doing a recreate with pulling the image by latest tag.
The best place to find the version of the image you are running is checking the Log on container startup.  

Regards,
André

**Reply by McTristan** [reply]
Actually I was wondering that too, now you are mentioning it. I thought the image versioning is based on the admins module main version. 
Why would it be a version 6 if there is no iobroker version 6?

**Reply by buanet** [reply]
You were wrong.
There is no connection between the admin adapter and the ioBroker Docker image. So why should the versions match???

Just for  clarification: ioBroker Docker image is NOT ioBroker. It is just an environment to run ioBroker. You can compare this a little to an operating system. The os brings everything you need to run ioBroker. The only connection between ioBroker software and the ioBroker Docker image is, that ioBroker is already preinstalled wenn you start it the first time. Like some apps are already preinstalled when you install Windows...

Regards,
André

**Reply by McTristan** [reply]
but the question remains: do I have to update iobroker always via a shell into the docker container?

**Reply by buanet** [reply]
Yes, like described in the docs: https://docs.buanet.de/iobroker-docker-image/docs/#updates-upgrades

**Comment by buanet** [comment]
@gyroslav Although it is a little outdated the update should work this way. I just changed the title and moved this to general discussions. 
I will go for adding this to the [official docs](https://docs.buanet.de). Thank you.

Regards,
 André

**Comment by TZanke** [comment]
Has anyone tested the new `maintenance script`  as a replacement for the solution described by @gyroslav ?

https://docs.buanet.de/iobroker-docker-image/docs/#use-maintenance-script-beta

I do not have any pending update at the moment, so I hopefully will find this comment next time I open the bookmark to this _How To_ if I need to.

Edit:
works!

Code:
```
maintenance on 
iobroker update
iobroker upgrade self
maintenance restart
```

Edit2:
see: https://docs.buanet.de/iobroker-docker-image/docs/#updates-upgrades

Code:
```
maintenance upgrade
```

**Comment by McR1983** [comment]


**Reply by buanet** [reply]
Steht aber eigentlich in der Meldung: 

![image](https://github.com/buanet/ioBroker.docker/assets/16137672/625503b4-e05b-4d88-9829-5fb3dca60dc1)

MfG,
André
